### ISSUE TITLE
Unable to run opkg update

### ISSUE THREAD

**Issue by kk246** [seud0nym/tch-gui-unhide#56]
Hi seud0nym,

I'm wondering if you can assist me.
I'm currently on firmware `18.1.c.0443-950-RB` and on the latest release of 
tch-gui-unhide (2021.12.06 Latest)

And I'm having a problem with running `opkg update`
I get the same result, whether through the console or via the web-ui, with error:
`wget: error getting response: Connection reset by peer`

Manually running `wget` in the console (even with `--no-check-certificate` option, still produces same results):
This only seems to happen with the domain `https://repository.macoers.com`
as downloading from `https://downloads.openwrt.org` seems to be ok.
(I am able to access the above two websites with no problem in my computer web browser)

I've tried a factory reset. and tried running `update-ca-certificates` to no avail.
Do you have any idea on what's going on?

Log details below
Thanks.

```
Fri Dec 24 11:25:23 2021 user.info rpc.gui.opkg: Downloading https://downloads.openwrt.org/releases/packages-21.02/arm_cortex-a9/base//ca-bundle_20210119-1_all.ipk...
Fri Dec 24 11:25:27 2021 user.info rpc.gui.opkg: Installing ca-bundle (20210119-1) to root...
Fri Dec 24 11:25:27 2021 user.info rpc.gui.opkg: Configuring ca-bundle.
Fri Dec 24 11:25:27 2021 user.info rpc.gui.opkg: Downloading https://downloads.openwrt.org/releases/packages-21.02/arm_cortex-a9/base//ca-certificates_20210119-1_all.ipk...
Fri Dec 24 11:25:30 2021 user.info rpc.gui.opkg: Installing ca-certificates (20210119-1) to root...
Fri Dec 24 11:25:30 2021 user.info rpc.gui.opkg: Configuring ca-certificates.
Fri Dec 24 11:25:30 2021 user.info rpc.gui.opkg: Updating ca-certificates.crt from https://curl.se/ca/cacert.pem if changed...
Fri Dec 24 11:28:45 2021 user.info rpc.gui.opkg: Package ca-bundle (20210119-1) installed in root is up to date.
Fri Dec 24 11:28:48 2021 user.info rpc.gui.opkg: Package ca-certificates (20210119-1) installed in root is up to date.
Fri Dec 24 11:28:57 2021 user.info rpc.gui.opkg: Executing: opkg update
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: wget: error getting response: Connection reset by peer
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: wget: error getting response: Connection reset by peer
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: wget: error getting response: Connection reset by peer
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: wget: error getting response: Connection reset by peer
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: wget: error getting response: Connection reset by peer
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: wget: error getting response: Connection reset by peer
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: Collected errors:
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg:  * opkg_download: Failed to download https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/base/Packages.gz, wget returned 1.
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg:  * opkg_download: Failed to download https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/packages/Packages.gz, wget returned 1.
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg:  * opkg_download: Failed to download https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/luci/Packages.gz, wget returned 1.
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg:  * opkg_download: Failed to download https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/routing/Packages.gz, wget returned 1.
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg:  * opkg_download: Failed to download https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/telephony/Packages.gz, wget returned 1.
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg:  * opkg_download: Failed to download https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/target/packages/Packages.gz, wget returned 1.
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/base/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: *** Failed to download the package list from https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/base/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/packages/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: *** Failed to download the package list from https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/packages/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/luci/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: *** Failed to download the package list from https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/luci/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/routing/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: *** Failed to download the package list from https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/routing/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/telephony/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: *** Failed to download the package list from https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/telephony/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/target/packages/Packages.gz
Fri Dec 24 11:28:58 2021 user.info rpc.gui.opkg: *** Failed to download the package list from https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/target/packages/Packages.gz

```

**Comment by seud0nym** [comment]
Very strange.

Are you able to access https://repository.macoers.com/ from your browser?

Are you using any adblocking like Pi-hole or AdGuard Home?

**Comment by kk246** [comment]
I definitely can access it via my web browser on my computer.
The DNS resolves to a proper IP Address.
So it's not a DNS blocker issue either.

```
root@DJA0231:~# wget "https://repository.macoers.com"
Connecting to repository.macoers.com (172.67.149.142:443)
wget: error getting response: Connection reset by peer
root@DJA0231:~#
```

**Comment by seud0nym** [comment]
A Google search turned up this issue: https://github.com/owtf/owtf/issues/592.  Same symptoms: curl and browser works, but wget fails when another firewall or proxy is involved... Could that be a cause?

i.e. does this work?
```
wget -U 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36 Edg/96.0.1054.62' https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/base/Packages.gz
```
Also, does
```
nslookup repository.macoers.com
```
return the same addresses on both your PC and your DJA0231?

**Comment by kk246** [comment]
No go unfortunately.

wget from DJA0231
```
root@DJA0231:/# wget -U 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari
/537.36 Edg/96.0.1054.62' https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/base/Packages.gz
Connecting to repository.macoers.com (172.67.149.142:443)
wget: error getting response: Connection reset by peer
root@DJA0231:/#
```

curl from DJA0231 (this works)
```
root@DJA0231:/# curl https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/base/Packages.gz -O
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 43946  100 43946    0     0  43946      0  0:00:01 --:--:--  0:00:01  137k
root@DJA0231:/# 
```

nslookup from DJA0231
```
root@DJA0231:/# nslookup repository.macoers.com
Server:    10.143.147.147
Address 1: 10.143.147.147

Name:      repository.macoers.com
Address 1: 172.67.149.142
Address 2: 104.21.29.174
root@DJA0231:/#
```

nslookup from windows 10 pc
```
C:\Users\username>nslookup repository.macoers.com
Server:  pi.hole
Address:  192.168.1.11

Non-authoritative answer:
Name:    repository.macoers.com
Addresses:  172.67.149.142
          104.21.29.174

C:\Users\username>

```

**Comment by seud0nym** [comment]
Well, I am confused.

What happens if you just use http: instead of https:?
```
wget http://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/base/Packages.gz
```

**Comment by kk246** [comment]
No change.
Http and https.
Same results.

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by seud0nym** [comment]
I have no idea what to suggest next.

Unfortunately wget on OpenWRT is not really wget. It is a cut-down version called uclient-fetch, and it doesn't have the same options for logging and debugging.

And AFAIK you can't replace wget in opkg.

The other problem is that since it is https (I have found that the repository is configured to redirect http: requests to https: which is why it doesn't make any difference), not even a tcpdump will help, because all the packets are encrypted.

Clutching at straws here... Maybe try setting the DNS server used by the DJA0231 to be your Pi-hole instead of 10.143.147.147? I don't know why it would make a difference, but it is the only thing that is different between the PC it works from and the DJA0231 that it doesn't.

**Comment by kk246** [comment]
I got `opkg update` completing successfully with a terrible wrapper for wget
opkg calls wget in this fashion.
`wget -q -O <output-file> <url>`


Contents of terrible wget wrapper below:
```
root@DJA0231:/tmp/11# cat /usr/bin/wget
#!/bin/sh
/usr/bin/curl -o "$3" "$4"

```

Log of opkg update using wget wrapper
```
root@DJA0231:/tmp/11# opkg update
Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/base/Packages.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 43946  100 43946    0     0  43946      0  0:00:01 --:--:--  0:00:01  130k
Updated list of available packages in /var/opkg-lists/base
Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/packages/Packages.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  227k  100  227k    0     0   227k      0  0:00:01 --:--:--  0:00:01  560k
Updated list of available packages in /var/opkg-lists/packages
Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/luci/Packages.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 49986  100 49986    0     0  49986      0  0:00:01 --:--:--  0:00:01  159k
Updated list of available packages in /var/opkg-lists/luci
Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/routing/Packages.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5557  100  5557    0     0   5557      0  0:00:01 --:--:--  0:00:01 22139
Updated list of available packages in /var/opkg-lists/routing
Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/telephony/Packages.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 70686  100 70686    0     0  70686      0  0:00:01 --:--:--  0:00:01  237k
Updated list of available packages in /var/opkg-lists/telephony
Downloading https://repository.macoers.com/homeware/18/brcm63xx-tch/VANTW/target/packages/Packages.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5266  100  5266    0     0   5266      0  0:00:01 --:--:--  0:00:01 18477
Updated list of available packages in /var/opkg-lists/core
root@DJA0231:/tmp/11#

```


A temporary work-around...
But I still don't know what the real problem is...

**Comment by kk246** [comment]
All good. 
I installed the `wget` opkg package from within the web-ui and I have the full fat `wget` command now
(Installing this package thankfully overwrites the wrapper i used above)

Rerunning the `opkg update` command, I have no problems.
Feel free to close this issue, if this fault is not reproducible on your end and you're not interested in investigating further

In summary:
- It must be something funky used with the wget inbuilt within the `18.1.c.0443-950-RB` firmware
- I know it was working before, but some recent change made it not work anymore
- The issue is not related to DNS level blocking
- The issue is not affected by either the http or https protocol
- It's not related to internet connectivity as by using the `curl` binary, was able to get around it

**Comment by seud0nym** [comment]
I am glad you found a solution, and I am sorry I wasn't more help.

**Comment by seud0nym** [comment]
I just encountered this problem when I restored 18.1.c to do some testing on a new script. It appears that in 18.1.c, wget is linked to the `busybox` binary, and not `uclient-fetch` as it is in 17.2 and 20.3.c. I couldn't find a better solution than your idea of applying the full fat version of `wget`, so I have now updated `de-telstra` and `tch-gui-unhide` to apply that on 18.1.c firmware.

Thanks again for coming up with the work-around.
